### PR TITLE
add neural ode optimization examples

### PR DESCRIPTION
**Pull request by jgoppert**
## Summary

Adds Neural ODE examples and initial optimization API scaffolding around a new `rumoca-opt` crate.

This PR includes:

- `rumoca-opt` as the canonical Rust API surface for model objectives, gradients, and gradient-descent style optimization experiments.
- Neural ODE user-guide documentation with tensor-native inference and native backprop/training examples.
- Example models and scenario configs for tensor Neural ODEs, native backprop, latent oscillator, and predator-prey dynamics.
- A reproducible `examples/benchmarks/neural_ode_jax_parity.py` benchmark that compares `NeuralODEBackprop1k` against an equivalent direct JAX implementation on CPU/GPU.
- Tensor MatMul lowering split out of `derivative_rhs.rs` and visible-observation lowering split out of `solve_model.rs` to stay under the SPEC_0021 file-size gate.
- Fixes for stale structured-equation metadata after derivative row reordering and after structural substitutions rewrite rows.
- Performance fixes for the 1k native backprop example: explicit `rk-like` value-only simulation lowering, indexed structural direct-definition lookup, batched algebraic assignment refresh, multi-output refresh plumbing, and structural prepare short-circuits for plain state-derivative models.
- A guarded direct RK-like lowering path for state-only ODE models. Models with algebraic/output solve slots, projected derivative dependencies, or state-selection constraints still fall back to full structural Modelica lowering.
- Regressions that check Neural ODE tensor lowering, time-dependent training batches, direct demotion, derivative expansion, guarded direct lowering, and the batched refresh path do not regress correctness.
- Fixes diffsol CI regressions from value-only SolveModel lowering by keeping Jacobian artifacts on the Auto/BDF path while leaving explicit `rk-like` on the value-only path.
- Fixes observation-time discrete refresh for `sample(time)`-derived rows so native simulation records event observations from refreshed values.
- Fixes no-state observation refresh so external table values settle before observation-time discrete rows are recorded.

## Why

The immediate goal is to demonstrate Neural ODE-style models in Rumoca using native tensor structure, then use those examples as the baseline for optimization and training API design.

The public API direction is intentionally conservative: `rumoca-opt` is introduced first, while deeper CLI/Python/WASM optimization adapters are left for follow-up PRs once the core Rust API survives more real examples.

The new direct lowering path is also conservative. It is first-class infrastructure for skipping structural work when the model is already a plain explicit state-only ODE. Neural ODE backprop models currently keep using structural lowering because their derivative rows depend on a large algebraic forward/backprop graph; accepting those directly regressed prepare and hot runtime until fused direct derivative lowering exists.

## Root Cause Fixed

The tensor lowering bug was stale structured-family metadata. Structured matrix/tensor equation families could be preserved through lowering, but later row reordering or substitution could change the row layout/body without updating the family metadata. Downstream scalarized tensor fallback could then read stale row-family metadata and emit invalid `LoadP` / `LoadIndexedP` accesses.

This PR remaps or drops structured families after derivative-row ordering, drops families whose rows are rewritten by substitution, and specifically covers residual rows (`lhs == None`) that previously returned before touched-row accounting.

The Neural ODE training parity bug was separate: `time` was not treated as a runtime dependency in the flatten/DAE parameter-variability classifiers. That let time-dependent training batch rows such as `sampleX` and target derivatives be cheapened/promoted as if constant, which made native gradients disagree with JAX. The classifiers now reject `time` as parameter-variable and flatten records terminal `time` references.

## Performance Notes

Benchmarked at 1,082 trainable parameters before scaling larger: `NeuralODEBackprop1k`, hidden width `H = 30`, `t_end = 0.2`, `dt = 0.02`, `iterations = 3`, `warmups = 1`, x64 JAX. Labels distinguish Rumoca's native Solve IR CPU simulator from the direct JAX equivalent used by the benchmark script.

| Metric | Rumoca native Solve IR CPU | JAX CPU equivalent | JAX GPU equivalent |
|---|---:|---:|---:|
| First usable integrated trajectory | 4.706 s `ready_seconds` | 0.471 s `integrate_compile_seconds` | 0.763 s `integrate_compile_seconds` |
| Hot integrated trajectory avg | 35.64 ms | 0.848 ms | 3.552 ms |
| Max parity error | 5.55e-17 vs JAX CPU | reference | 8.88e-16 vs Rumoca/JAX CPU |

Current Rumoca native CPU first-run breakdown for this model:

- `compile_seconds = 0.214`
- `prepare_seconds = 4.493`
- `ready_seconds = 4.706`
- `prepare_ir_solve_structural_dae_seconds = 3.473`
- `prepare_ir_solve_lower_seconds = 0.985`
- `prepare_backend_build_seconds = 0.018`

This is not speed parity yet. Current first-run wait is `9.99x` JAX CPU integrated compile time and `6.17x` JAX GPU integrated compile time. Current hot trajectory time is `42.03x` JAX CPU and `10.03x` JAX GPU. The remaining native gap is mostly structural DAE/Solve preparation plus scalar hot-loop execution of the backprop equations; the benchmark records those buckets directly so follow-up work can track progress without hand timing.

Accuracy was checked in the benchmark, not assumed: initial state, initial observables, initial gradients, final state, final observables, and JAX manual backprop versus `jax.value_and_grad` all match at roundoff-level error. Rumoca is executing the explicit Modelica backpropagation equations through the native Solve IR CPU path; JAX is using `jax.value_and_grad` for the equivalent RHS.

## Validation

Latest checks run after the guarded direct-lowering update:

- `cargo fmt --check`
- `cargo check -p rumoca-sim -p rumoca`
- `cargo test -p rumoca-sim solve_lowering -- --nocapture`
- `cargo run --release -p rumoca -- sim bench examples/models/NeuralODEBackprop.mo --model NeuralODEBackprop1k --solver rk-like --t-end 0.2 --dt 0.02 --iterations 3 --warmups 1 --json`

Earlier checks on this PR also include:

- `cargo check -p rumoca-phase-structural -p rumoca-phase-flatten -p rumoca-phase-dae -p rumoca-phase-solve -p rumoca-sim -p rumoca`
- `cargo test -p rumoca-phase-flatten param_variability::tests`
- `cargo test -p rumoca-phase-dae time_is_not_parameter_variable_reference`
- `cargo test -p rumoca-phase-structural demote_direct_assigned_states`
- `cargo test -p rumoca-phase-structural compound_derivative_expansion`
- `cargo test -p rumoca-phase-solve variable_meta`
- `cargo test -p rumoca-phase-solve derivative_row`
- `python3 -m py_compile examples/benchmarks/neural_ode_jax_parity.py`
- `target/jax-bench-venv/bin/python examples/benchmarks/neural_ode_jax_parity.py --platform cpu --json-output target/neural_ode_jax_parity_cpu.json`
- `target/jax-bench-venv/bin/python examples/benchmarks/neural_ode_jax_parity.py --platform gpu --json-output target/neural_ode_jax_parity_gpu.json`

Earlier in this branch, before the final benchmark/performance updates, `cargo xtask verify full` also passed.

### CODE REVIEW

**Comment by jgoppert** [comment]
Closed in favor of #291
